spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TimeDistributedMaskCriterion.scala

```diff
@@ -0,0 +1,160 @@
+import scala.concurrent.Future
+import scala.reflect.ClassTag
+
+/**
+ * This class is intended to support inputs with 3 or more dimensions.
+ * Apply Any Provided Criterion to every temporal slice of an input.
+ * In addition, it supports padding mask.
+ *
+ * eg. if the target is [ [-1, 1, 2, 3, -1], [5, 4, 3, -1, -1] ],
+ *     and set the paddingValue property to -1, then the loss of -1 would not
+ *     be accumulated and the loss is only divided by 6 (ont including the amount of
+ *     -1, in this case, we are only interested in 1, 2, 3, 5, 4, 3)
+ *
+ * @param critrn embedded criterion
+ * @param dimension to compute criterion loss along the input and target dimension
+ */
+
+class TimeDistributedMaskCriterion[T : ClassTag](
+  val critrn : TensorCriterion[T],
+  val dimension: Int = 2,
+  val paddingValue: Int = 0
+)
+  (implicit ev: TensorNumeric[T]) extends TensorCriterion[T] {
+
+  private var fInput: Tensor[T] = Tensor[T]()
+  private var fTarget: Tensor[T] = Tensor[T]()
+  private var _gradInput = Tensor[T]()  // list of cell criterions cloned from added criterion
+  private val cells: ArrayBuffer[TensorCriterion[T]]
+  = ArrayBuffer[TensorCriterion[T]]()
+
+  @transient
+  protected var results: Array[Future[Unit]] = _
+  private val mask = Tensor[T]()
+
+  /**
+   * Clone N criterions; N depends on the time dimension of the input
+   * @param times
+   */
+  private def extend(times: Int): Unit = {
+    var t = cells.length
+    while (t < times) {
+      cells += critrn.cloneCriterion()
+        .asInstanceOf[TensorCriterion[T]]
+      t += 1
+    }
+  }
+
+
+  override def updateOutput(input: Tensor[T], target: Tensor[T]): T = {
+    /**
+     * Take each time slice of input and target, and add up all outputs of slices
+     * Example with dimension=2:
+     * input.size = [B, T, D] => fInput.size = [B, D]
+     * target.size = [B, T] => fTarget.size = [B]
+     */
+    require(input.size(dimension) == target.size(dimension),
+      "target should have as many elements as input, " +
+        s"input ${input.size(dimension)}, target ${target.size(dimension)}")
+
+    output = ev.fromType[Int](0)
+    val nstep = input.size(dimension)
+    extend(nstep)
+
+    if (results == null || results.length != nstep) {
+      results = new Array[Future[Unit]](nstep)
+    }
+
+    var i = 0
+    while (i < nstep) {
+      val _i = i + 1
+      results(i) = Engine.model.invoke(() => {
+        fInput = input.select(dimension, _i)
+        fTarget = target.select(dimension, _i)
+        cells(_i - 1).updateOutput(fInput, fTarget)
+      })
+      i += 1
+    }
+    Engine.model.sync(results)
+
+    mask.resizeAs(target)
+    mask.applyFun[T](target, x =>
+      if (x != ev.fromType[Int](paddingValue)) ev.one else ev.zero)
+    val outputBuff = Tensor()
+    outputBuff.resizeAs(mask)
+    (0 until nstep).foreach(b => {
+      outputBuff
+        .select(2, b + 1)
+        .cmul(mask.select(2, b + 1),
+          cells(b).sparseOutput)
+    })
```

Comment:
I think sparseOutput is not needed. 
You can define getSizeAverage in criterion which can determine the compute mode.